olm-catalog/serverless-operator-index/Dockerfile

```diff
@@ -9,13 +9,9 @@ COPY olm-catalog/serverless-operator-index/configs /configs
 
 RUN /bin/opm init serverless-operator --default-channel=stable --output yaml >> /configs/index.yaml
 RUN /bin/opm render --skip-tls-verify -o yaml \
-registry.ci.openshift.org/knative/release-1.33.0:serverless-bundle \
 registry.ci.openshift.org/knative/release-1.34.0:serverless-bundle \
-      registry.ci.openshift.org/knative/release-1.35.0:serverless-bundle >> /configs/index.yaml || \
-    /bin/opm render --skip-tls-verify -o yaml \
-registry.ci.openshift.org/knative/release-1.33.0:serverless-bundle \
-registry.ci.openshift.org/knative/release-1.34.0:serverless-bundle \
-      quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-135/serverless-bundle@sha256:0d5feb3c9f59246d16db1c687e3c1085fcc8a360cd51974197a878944df06b1d >> /configs/index.yaml
+registry.ci.openshift.org/knative/release-1.35.0:serverless-bundle \
+       >> /configs/index.yaml
```

Comment:
@mgencur this is the issue `SERVERLESS_BUNDLE` in `dockerfile.sh` is empty

Reply:
```
++++ skopeo inspect --no-tags=true docker://quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle:latest
++++ jq -r .Digest
FATA[0001] Error parsing image name "docker://quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle:latest": reading manifest latest in quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle: unauthorized: access to the requested resource is not authorized 
++++ echo ''
+++ digest=
+++ '[' '' = '' ']'
+++ '[' false = true ']'
+++ echo ''
+++ return
++ image=
++ [[ '' == '' ]]
+++ image_with_sha registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0
+++ image=registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0
+++ return_input_on_empty=false
++++ skopeo inspect --no-tags=true docker://registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0
++++ jq -r .Digest
FATA[0000] Error parsing image name "docker://registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0": reading manifest release-1.36.0 in registry.ci.openshift.org/knative/serverless-bundle: manifest unknown 
```

it can't find either 
`quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle`
`registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0`

(expected)

Reply:
I'm wondering if we need to also fallback to `knative-main` in `get_bundle_for_version` when not even CI `registry.ci.openshift.org/knative/serverless-bundle:release-1.36.0` is available?

```
  # As a backup, try also CI registry.
  if [[ "${image}" == "" ]]; then
    image=$(image_with_sha "registry.ci.openshift.org/knative/serverless-bundle:release-${version}")
  fi
  if [[ "${image}" == "" ]]; then
    image=$(image_with_sha "registry.ci.openshift.org/knative/serverless-bundle:knative-main")
  fi

```

Reply:
This will fix it for the Dockerfile: https://gist.github.com/mgencur/5316c4655abcb92e28a3455ee02a0b5f
But I'm still checking if the final catalog can be generated properly if the latest bundle doesn't exist yet.

Reply:
My solution can't build the catalog properly because it requires the image to exist:
```
INFO    16:41:10.114 Generating catalog for OCP 4.14
INFO[0000] rendering index "registry.redhat.io/redhat/redhat-operator-index:v4.14" as file-based catalog 
INFO[0404] wrote rendered file-based catalog to "/tmp/knative.JC1YvFt4/tmp.hVU6K2GqdC" 
FATA[0002] Error parsing image name "docker://quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle:latest": reading manifest latest in quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle: unauthorized: access to the requested resource is not authorized 
2024/11/27 16:49:45 render reference "quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle:latest": error resolving name for image ref quay.io/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle:latest: unexpected status from HEAD request to https://quay.io/v2/redhat-user-workloads/ocp-serverless-tenant/serverless-operator-136/serverless-bundle/manifests/latest: 401 UNAUTHORIZED
ERROR   16:49:45.652 🚨 Error (code: 1) occurred at ./hack/generate/catalog.sh:49, with command: opm alpha render-template basic --migrate-level="$level" "${catalog_template}" -oyaml > "${index_dir}/v${ocp_version}/catalog/serverless-operator/catalog.yaml"

```

Reply:
So, yeah. We could fall back to registry.ci.openshift.org/knative/serverless-bundle:knative-main